### PR TITLE
update doc

### PR DESCRIPTION
**Pull request by monperrus**

### CODE REVIEW

File: doc/filter.md [hunk 1]
```diff
@@ -47,27 +47,33 @@ Queries
...
     .filterChildren(new AnnotationFilter(Deprecated.class))
-    .filterChildren(new TypeFilter(CtMethod.class)).list();
 ```
 
-`CtQueryable#map(CtFunction)`enables you to give Java 8 lambda as query.
 A boolean return value of the lambda tells whether the elements are selected for inclusion or not.
 
 ```java
 // creating a custom filter to select all public fields using java 8 lambda
 list3 = rootPackage.filterChildren((CtField field)->field.getModifiers.contains(ModifierKind.PUBLIC)).list();
```
Comment (pvojtechovsky) [inline doc/filter.md:73]: The examples above does not evaluate query using `list` or `forEach`. Only this example ends with `list()`. Is it intentional?

File: src/main/java/spoon/reflect/visitor/chain/CtQuery.java [hunk 2]
```diff
@@ -23,59 +23,91 @@
 
 /**
  * CtQuery represents a query, which can be used to traverse a spoon model and collect
- * children elements in several ways.<br>
+ * children elements in several ways.<br/>
  *
- * <br>
- * Use {@link CtQueryable#map(CtFunction)}} or {@link CtQueryable#filterChildren(Filter)} to create a new query starting from an existing element.<br>
+ * Creation: A query is created either from a {@link CtElement}, or it can be used defined first from {@link CtQueryImpl} and bound to root elements
```
Comment (pvojtechovsky) [inline src/main/java/spoon/reflect/visitor/chain/CtQuery.java:28]: "can be used defined first" ?? May be delete "used"?

File: src/main/java/spoon/reflect/visitor/chain/CtQuery.java [hunk 3]
```diff
@@ -23,59 +23,91 @@
...
+	 * The matched child element for which (filter.matches(element)==true) are sent to the next query step.
+	 * Essentially the same as {@link CtElement#getElements(Filter)} but more powerful, because it
+	 * can be chained with other subsequent queries.
+	 *
+	 * Note: the input element (the root of the query, `this` if you're in {@link CtElement}) is also checked and may thus be also sent to the next step.
+	 * The elements which throw {@link ClassCastException} during {@link Filter#matches(CtElement)}
+	 * are considered as **not matching**, ie. are excluded.
+	 *
+	 * @param filter used to filter scanned children elements of the AST tree
+	 * @return a new Query
```
Comment (pvojtechovsky) [inline src/main/java/spoon/reflect/visitor/chain/CtQuery.java:50]: just note... do not know if relevant ...
it is "a new Query" only when called first time on CtElement. Later the same query instance is returned (this)

File: src/main/java/spoon/reflect/visitor/chain/CtQuery.java [hunk 4]
```diff
@@ -23,59 +23,91 @@
...
+	 */
+	@Override
+	<R extends CtElement> CtQuery filterChildren(Filter<R> filter);
+
+
+	/**
+	 * Query elements based on a function, the behavior depends on the return type of the function.
+	 * <table>
+	 * <tr><td><b>Return type of `function`</b><td><b>Behavior</b>
+	 * <tr><td>{@link Boolean}<td>Select elements if thereturned value of `function` is true (as for {@link Filter}).
```
Comment (pvojtechovsky) [inline src/main/java/spoon/reflect/visitor/chain/CtQuery.java:60]: "thereturned"->"the returned"

File: src/main/java/spoon/reflect/visitor/chain/CtQuery.java [hunk 5]
```diff
@@ -23,59 +23,91 @@
...
+	 * </table><br>
+	 *
+	 * @param function a Function with one parameter of type I returning a value of type R
+	 * @return a new query object
+	 */
+	@Override
+	<I, R> CtQuery map(CtFunction<I, R> function);
+
+	/**
+	 * Sets (binds) the input of the query, the search will be done on children of that element.
```
Comment (pvojtechovsky) [inline src/main/java/spoon/reflect/visitor/chain/CtQuery.java:73]: the search will be done **on children** of that element - it is true only for filterChildren... but how to write it better? :-)

File: src/main/java/spoon/reflect/visitor/chain/CtQuery.java [hunk 6]
```diff
@@ -23,59 +23,91 @@
...
+	 * If the query is created by {@link CtElement#map} or {@link CtElement#filterChildren(Filter)},
 	 * then the query is already bound to this element.
-	 * A new call of {@link #setInput(Object...)} will reset the current binding ans use the new one.
+	 * A new call of {@link #setInput(Object...)} will is always possible, it resets the current binding ans sets the new one.
```
Comment (pvojtechovsky) [inline src/main/java/spoon/reflect/visitor/chain/CtQuery.java:76]: " will is always "->  "is always "

File: src/main/java/spoon/reflect/visitor/chain/CtQuery.java [hunk 7]
```diff
@@ -23,59 +23,91 @@
...
 	 *
 	 * @param input
 	 * @return this to support fluent API
 	 */
 	<T extends CtQuery> T setInput(Object... input);
 
 	/**
-	 * actually evaluates the query and for each produced outputElement calls `consumer.accept(outputElement)`
+	 * Actually evaluates the query and for each produced output element of the last step,
```
Comment (pvojtechovsky) [inline src/main/java/spoon/reflect/visitor/chain/CtQuery.java:84]: > evaluates the query and for each produced output element
remove "and"

File: src/main/java/spoon/reflect/visitor/chain/CtQuery.java [hunk 8]
```diff
@@ -23,59 +23,91 @@
...
 	<R> void forEach(CtConsumer<R> consumer);
+
 	/**
-	 * actually evaluates the query and returns all the produced elements collected in a List
+	 * Actually evaluates the query and returns all the elements produced in the last step.
 	 * @return the list of elements collected by the query.
 	 * @see #forEach(CtConsumer) for an efficient way of manipulating the elements without creating an intermediate list.
 	 */
 	<R extends Object> List<R> list();
+
```
Comment (pvojtechovsky) [inline src/main/java/spoon/reflect/visitor/chain/CtQuery.java:99]: What about note like this:
You can use any item type of the List. The compiler nor the query engine will check it. So the wrong chosen List type can cause ClassCastException

File: src/main/java/spoon/reflect/visitor/chain/CtQuery.java [hunk 9]
```diff
@@ -23,59 +23,91 @@
...
 	/**
-	 * actually evaluates the query and returns these produced elements as a List,
-	 * which are assignable to `itemClass`
+	 * Same as {@link #list()}, but with static typing on the return type.
```
Comment (pvojtechovsky) [inline src/main/java/spoon/reflect/visitor/chain/CtQuery.java:101]: but with static typing on the return type and the **final filtering**, which matches only results which are assignable from that return type.

File: src/main/java/spoon/reflect/visitor/chain/CtQuery.java (hunk 3 above)
Comment (pvojtechovsky) [reply src/main/java/spoon/reflect/visitor/chain/CtQuery.java:50]: On other places there is:
>  @return this to support fluent API

**Comment by pvojtechovsky** [comment]
I am done with review of that excellent documentation.

**Comment by monperrus** [comment]
Thanks. For this time, and esp for the next ones to come, it would be much more efficient that you do those last changes yourself:

* a PR on my PR on your PR (overkill?) :-)
* you merge and fix the last typos

**Pull request merged**